app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt

```diff
@@ -0,0 +1,247 @@
+        serviceTable.onItemClickListener = OnItemClickListener { parent: AdapterView<*>, view: View?, position: Int, id: Long ->
+            val info = parent.getItemAtPosition(position) as ServiceInfoWrapper
+            connectToChild(info.address, info.port, info.name)
+        }
+
+        // Instantiate a new DiscoveryListener
+        discoveryListener = object : DiscoveryListener {
+            //  Called as soon as service discovery begins.
+            override fun onDiscoveryStarted(regType: String) {
+                Log.d(TAG, "Service discovery started")
+            }
+
+            override fun onServiceFound(service: NsdServiceInfo) {
+                // A service was found!  Do something with it.
+                Log.d(TAG, "Service discovery success: $service")
+                if (service.serviceType != serviceType) {
+                    // Service type is the string containing the protocol and
+                    // transport layer for this service.
+                    Log.d(TAG, "Unknown Service Type: " + service.serviceType)
+                } else if (service.serviceName.contains("ChildMonitor")) {
+                    val resolver: NsdManager.ResolveListener = object : NsdManager.ResolveListener {
+                        override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
+                            // Called when the resolve fails.  Use the error code to debug.
+                            Log.e(TAG, "Resolve failed: error $errorCode for service: $serviceInfo")
+                        }
+
+                        override fun onServiceResolved(serviceInfo: NsdServiceInfo) {
+                            Log.i(TAG, "Resolve Succeeded: $serviceInfo")
+                            runOnUiThread {
+                                for (index in 0 until availableServicesAdapter.count) {
+                                    val item = availableServicesAdapter.getItem(index)
+                                    if (item != null && item.matches(serviceInfo)) {
+                                        // Prevent inserting duplicates
+                                        return@runOnUiThread
+                                    }
+                                }
+                                availableServicesAdapter.add(ServiceInfoWrapper(serviceInfo))
+                            }
+                        }
+                    }
+                    this@DiscoverActivity.nsdManager!!.resolveService(service, resolver)
+                } else {
+                    Log.d(TAG, "Unknown Service name: " + service.serviceName)
+                }
+            }
+
+            override fun onServiceLost(service: NsdServiceInfo) {
+                // When the network service is no longer available.
+                // Internal bookkeeping code goes here.
+                Log.e(TAG, "Service lost: $service")
+                multicastReleaser.run()
```

Comment:
One this is no longer a Runnable you'll be able to just call it, instead of calling run()

Reply:
It turns into `multicastReleaser?.invoke()`  😄